File: src/models/Listing.ts
import mongoose, { Schema, Document, Types } from 'mongoose';

export interface IListing extends Document {
  title: string;
  description: string;
  category: string;
  price: number;
  originalPrice?: number;
  condition: string;
  fabric: string;
  size: {
    chest?: number;
    waist?: number;
    hip?: number;
    length?: number;
    shoulder?: number;
  };
  images: string[];
  defects?: string;
  city: string;
  seller: Types.ObjectId;
  featured: boolean;
  isFeatured?: boolean;
  featuredUntil?: Date;
  views: number;
  status: 'active' | 'paused' | 'sold';
  createdAt: Date;
  updatedAt: Date;
}

const listingSchema = new Schema<IListing>(
  {
    title: {
      type: String,
      required: [true, 'Title is required'],
      trim: true,
      minlength: [5, 'Title must be at least 5 characters'],
      maxlength: [100, 'Title cannot exceed 100 characters'],
    },
    description: {
      type: String,
      required: [true, 'Description is required'],
      minlength: [20, 'Description must be at least 20 characters'],
      maxlength: [2000, 'Description cannot exceed 2000 characters'],
    },
    category: {
      type: String,
      required: [true, 'Category is required'],
      enum: ['bridal', 'groom-wear', 'formal-party', 'casual-ethnic', 'kids', 'accessories'],
    },
    price: {
      type: Number,
      required: [true, 'Price is required'],
      min: [100, 'Price must be at least 100'],
    },
    originalPrice: {
      type: Number,
      default: null,
    },
    condition: {
      type: String,
      required: [true, 'Condition is required'],
      enum: ['new_with_tags', 'worn_once', 'worn_few_times', 'minor_alterations'],
    },
    fabric: {
      type: String,
      required: [true, 'Fabric type is required'],
      enum: [
        'Silk',
        'Chiffon',
        'Banarsi',
        'Organza',
        'Velvet',
        'Georgette',
        'Net',
        'Cotton',
        'Lawn',
        'Khaddar',
        'Jamawar',
      ],
    },
    size: {
      chest: { type: Number, default: null },
      waist: { type: Number, default: null },
      hip: { type: Number, default: null },
      length: { type: Number, default: null },
      shoulder: { type: Number, default: null },
    },
    images: {
      type: [String],
      required: [true, 'At least one image is required'],
      validate: {
        validator: function (v: string[]) {
          return v && v.length >= 1 && v.length <= 10;
        },
        message: 'Images must be between 1 and 10',
      },
    },
    defects: {
      type: String,
      default: null,
    },
    city: {
      type: String,
      required: [true, 'City is required'],
    },
    seller: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      required: [true, 'Seller is required'],
    },
    featured: {
      type: Boolean,
      default: false,
    },
    isFeatured: {
      type: Boolean,
      default: false,
    },
    featuredUntil: {
      type: Date,
      default: null,
    },
    views: {
      type: Number,
      default: 0,
    },
    status: {
      type: String,
      enum: ['active', 'paused', 'sold'],
      default: 'active',
    },
  },
  { timestamps: true }
);

// Index for fast queries
listingSchema.index({ category: 1, city: 1, status: 1 });
listingSchema.index({ seller: 1 });
listingSchema.index({ createdAt: -1 });
listingSchema.index({ featured: 1, status: 1 });

export default mongoose.models.Listing || mongoose.model<IListing>('Listing', listingSchema);

File: flutter_app/lib/models/listing.dart
class Listing {
  final String id;
  final String title;
  final String description;
  final String category;
  final int price;
  final int? originalPrice;
  final String condition;
  final String fabric;
  final Map<String, dynamic> size;
  final List<String> images;
  final String? defects;
  final String city;
  final String sellerId;
  final String? sellerName;
  final bool featured;
  final bool? isFeatured;
  final DateTime? featuredUntil;
  final int views;
  final String status;
  final DateTime createdAt;

  Listing({
    required this.id,
    required this.title,
    required this.description,
    required this.category,
    required this.price,
    this.originalPrice,
    required this.condition,
    required this.fabric,
    required this.size,
    required this.images,
    this.defects,
    required this.city,
    required this.sellerId,
    this.sellerName,
    this.featured = false,
    this.isFeatured = false,
    this.featuredUntil,
    this.views = 0,
    this.status = 'active',
    required this.createdAt,
  });

  factory Listing.fromJson(Map<String, dynamic> json) {
    return Listing(
      id: json['_id'] ?? '',
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      category: json['category'] ?? '',
      price: json['price'] ?? 0,
      originalPrice: json['originalPrice'],
      condition: json['condition'] ?? '',
      fabric: json['fabric'] ?? '',
      size: json['size'] ?? {},
      images: List<String>.from(json['images'] ?? []),
      defects: json['defects'],
      city: json['city'] ?? '',
      sellerId: json['seller'] is Map ? json['seller']['_id'] : json['seller'] ?? '',
      sellerName: json['seller'] is Map ? json['seller']['name'] : null,
      featured: json['featured'] ?? false,
      isFeatured: json['isFeatured'] ?? false,
      featuredUntil: json['featuredUntil'] != null ? DateTime.parse(json['featuredUntil']) : null,
      views: json['views'] ?? 0,
      status: json['status'] ?? 'active',
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toString()),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'title': title,
      'description': description,
      'category': category,
      'price': price,
      'originalPrice': originalPrice,
      'condition': condition,
      'fabric': fabric,
      'size': size,
      'images': images,
      'defects': defects,
      'city': city,
      'seller': sellerId,
      'featured': featured,
      'isFeatured': isFeatured,
      'featuredUntil': featuredUntil?.toIso8601String(),
      'views': views,
      'status': status,
      'createdAt': createdAt.toIso8601String(),
    };
  }

  int get discount {
    if (originalPrice == null || originalPrice == 0) return 0;
    return (((originalPrice! - price) / originalPrice!) * 100).toInt();
  }

  bool get isSold => status == 'sold';
  bool get isPaused => status == 'paused';
  bool get isActive => status == 'active';

  Listing copyWith({
    String? id,
    String? title,
    String? description,
    String? category,
    int? price,
    int? originalPrice,
    String? condition,
    String? fabric,
    Map<String, dynamic>? size,
    List<String>? images,
    String? defects,
    String? city,
    String? sellerId,
    String? sellerName,
    bool? featured,
    bool? isFeatured,
    DateTime? featuredUntil,
    int? views,
    String? status,
    DateTime? createdAt,
  }) {
    return Listing(
      id: id ?? this.id,
      title: title ?? this.title,
      description: description ?? this.description,
      category: category ?? this.category,
      price: price ?? this.price,
      originalPrice: originalPrice ?? this.originalPrice,
      condition: condition ?? this.condition,
      fabric: fabric ?? this.fabric,
      size: size ?? this.size,
      images: images ?? this.images,
      defects: defects ?? this.defects,
      city: city ?? this.city,
      sellerId: sellerId ?? this.sellerId,
      sellerName: sellerName ?? this.sellerName,
      featured: featured ?? this.featured,
      isFeatured: isFeatured ?? this.isFeatured,
      featuredUntil: featuredUntil ?? this.featuredUntil,
      views: views ?? this.views,
      status: status ?? this.status,
      createdAt: createdAt ?? this.createdAt,
    );
  }
}
